### PR TITLE
Fix deadlock on BKSyncLogReaderDLSN

### PR DESCRIPTION
**Pull request by khurrumnasimm**
Change the lastSeenDLSN to volatile to remove the synchronization block to avoid deadlock with sharedLock

### CODE REVIEW

File: distributedlog-core/src/main/java/com/twitter/distributedlog/BKSyncLogReaderDLSN.java [hunk 1]
```diff
@@ -116,7 +110,7 @@ public void resumeReadAhead() {
 
     @Override
     public void onSuccess(LogRecordWithDLSN record) {
-        setLastSeenDLSN(record.getDlsn());
+        this.lastSeenDLSN = record.getDlsn();
```
Comment (leighst) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/BKSyncLogReaderDLSN.java]: lastSeenDLSN is referenced below in a multi statement block - are you certain atomicity isn't required there?

Comment (khurrumnasimm) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/BKSyncLogReaderDLSN.java]: Are you referring this code below? lastDLSNSeenByReadAhead is a local variable and the lastSeenDLSN is a volatile variable. The visibility and the ordering of a volatile variable is guaranteed by java - that says accessing the lastSeenDLSN will retrieve the latest value, hence no atomicity is required here.

`                        // if last seen DLSN by reader is same as the one seen by ReadAhead
                        // that means that reader is caught up with ReadAhead and ReadAhead
                        // is caught up with stream
                        shallWait = DLSN.InitialDLSN != lastDLSNSeenByReadAhead
                                && lastSeenDLSN.compareTo(lastDLSNSeenByReadAhead) < 0
                                && startDLSN.compareTo(lastDLSNSeenByReadAhead) <= 0;`

Comment (khurrumnasimm) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/BKSyncLogReaderDLSN.java]: I agreed. I don't think we need "synchronized (sharedLock)" block.

Comment (khurrumnasimm) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/BKSyncLogReaderDLSN.java]: @leighst I pushed another change to remove 'synchronized' block. please review it again. thank you in advance.

Comment (sijie) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/BKSyncLogReaderDLSN.java]: LGTM. @leighst what do you think?

Comment (leighst) [inline distributedlog-core/src/main/java/com/twitter/distributedlog/BKSyncLogReaderDLSN.java]: shipit

**Pull request closed without merging**